src/lib/consensus/proof_of_stake.ml

```diff
@@ -1030,35 +1045,105 @@ module Make (Inputs : Inputs_intf) :
       ( transition |> Snark_transition.blockchain_state
       |> Blockchain_state.ledger_hash )
 
-  let select this that ~logger ~time_received =
+  let select old_state new_state ~logger ~time_received =
     let open Consensus_state in
     let open Epoch_data in
     let logger = Logger.child logger "proof_of_stake" in
+    let string_of_choice = function `Take -> "Take" | `Keep -> "Keep" in
+    let log_result choice msg =
+      Logger.debug logger "RESULT: %s -- %s" (string_of_choice choice) msg
+    in
+    let log_choice ~precondition_msg ~choice_msg choice =
+      let choice_msg =
+        match choice with
+        | `Take -> choice_msg
+        | `Keep -> Printf.sprintf "not (%s)" choice_msg
+      in
+      let msg = Printf.sprintf "(%s) && (%s)" precondition_msg choice_msg in
+      log_result choice msg
+    in
+    Logger.info logger "SELECTING BEST CONSENSUS STATE" ;
+    Logger.info logger
+      !"old consensus state: %{sexp:Consensus_state.value}"
+      old_state ;
+    Logger.info logger
+      !"new consensus state: %{sexp:Consensus_state.value}"
+      new_state ;
     (* TODO: update time_received check and `Keep when it is not met *)
     if
       not
-        (time_in_epoch_slot that
+        (time_in_epoch_slot new_state
            Time.(of_span_since_epoch (Span.of_ms time_received)))
     then Logger.error logger "received a transition outside of it's slot time" ;
     (* TODO: add fork_before_checkpoint check *)
-    (* not (List.exists this.checkpoints ~f:(fun c ->
-     *   List.exists that.checkpoints ~f:(
-     *     Coda_base.State_hash.equal c)))
-     *)
-    if
-      Coda_base.State_hash.equal this.last_epoch_data.lock_checkpoint
-        that.last_epoch_data.lock_checkpoint
-    then if Length.compare this.length that.length < 0 then `Take else `Keep
-    else if
-      Coda_base.State_hash.equal this.last_epoch_data.start_checkpoint
-        that.last_epoch_data.start_checkpoint
-    then
-      if
-        Length.compare this.last_epoch_data.length that.last_epoch_data.length
-        < 0
-      then `Take
-      else `Keep
-    else `Keep
+    (* Each branch contains a precondition predicate and a choice predicate,
+     * which takes the new state when true. Each predicate is also decorated
+     * with a string description, used for debugging messages *)
+    let ( = ) = Coda_base.State_hash.equal in
+    let ( < ) a b = Length.compare a b < 0 in
+    let branches =
```

Comment:
Maybe we can try and reduce the cases?
If we store the most recent start_checkpoints for both and we store the most recent lock_checkpoints for both, I think we can just compare directly, so we don't need all these different branch cases. I think that makes sense, grab me in person and we can discuss if not

Reply:
I split it up like this for debugging for now. It would be nice to keep it split out for a little while longer while I'm still debugging and stabilizing proof of stake. Should I fix it now or make an issue for it?

Reply:
Was thinking fix it now? Not sure if what I was saying above makes full sense, but as I'm visualizing it would end up really simple, and be pretty obvious its bug free